File: src/Systems/Go/builder.go
package main

import (
	"fmt"
	"strings"
)

type ReportBuilder interface {
	Reset()
	AddTitle(string)
	AddSection(string, string)
	Build() string
}

type TextReportBuilder struct {
	parts []string
}

func (b *TextReportBuilder) Reset() {
	b.parts = nil
}

func (b *TextReportBuilder) AddTitle(title string) {
	b.parts = append(b.parts, "# "+title)
}

func (b *TextReportBuilder) AddSection(heading, body string) {
	b.parts = append(b.parts, "## "+heading, body)
}

func (b *TextReportBuilder) Build() string {
	return strings.Join(b.parts, "\n")
}

type HTMLReportBuilder struct {
	parts []string
}

func (b *HTMLReportBuilder) Reset() {
	b.parts = nil
}

func (b *HTMLReportBuilder) AddTitle(title string) {
	b.parts = append(b.parts, "<h1>"+title+"</h1>")
}

func (b *HTMLReportBuilder) AddSection(heading, body string) {
	b.parts = append(b.parts, "<h2>"+heading+"</h2>", "<p>"+body+"</p>")
}

func (b *HTMLReportBuilder) Build() string {
	return strings.Join(b.parts, "")
}

func BuildAvailabilityReport(builder ReportBuilder) string {
	builder.Reset()
	builder.AddTitle("Service status")
	builder.AddSection("Availability", "99.95%")
	return builder.Build()
}

func main() {
	fmt.Println(BuildAvailabilityReport(&TextReportBuilder{}))
	fmt.Println("---")
	fmt.Println(BuildAvailabilityReport(&HTMLReportBuilder{}))
}

File: src/Systems/Rust/builder.rs
trait ReportBuilder {
    fn reset(&mut self);
    fn add_title(&mut self, title: &str);
    fn add_section(&mut self, heading: &str, body: &str);
    fn build(&self) -> String;
}

#[derive(Default)]
struct TextReportBuilder {
    parts: Vec<String>,
}

impl ReportBuilder for TextReportBuilder {
    fn reset(&mut self) {
        self.parts.clear();
    }

    fn add_title(&mut self, title: &str) {
        self.parts.push(format!("# {title}"));
    }

    fn add_section(&mut self, heading: &str, body: &str) {
        self.parts.push(format!("## {heading}"));
        self.parts.push(body.to_owned());
    }

    fn build(&self) -> String {
        self.parts.join("\n")
    }
}

#[derive(Default)]
struct HtmlReportBuilder {
    parts: Vec<String>,
}

impl ReportBuilder for HtmlReportBuilder {
    fn reset(&mut self) {
        self.parts.clear();
    }

    fn add_title(&mut self, title: &str) {
        self.parts.push(format!("<h1>{title}</h1>"));
    }

    fn add_section(&mut self, heading: &str, body: &str) {
        self.parts.push(format!("<h2>{heading}</h2>"));
        self.parts.push(format!("<p>{body}</p>"));
    }

    fn build(&self) -> String {
        self.parts.concat()
    }
}

fn build_availability_report(builder: &mut dyn ReportBuilder) -> String {
    builder.reset();
    builder.add_title("Service status");
    builder.add_section("Availability", "99.95%");
    builder.build()
}

fn main() {
    println!(
        "{}",
        build_availability_report(&mut TextReportBuilder::default())
    );
    println!("---");
    println!(
        "{}",
        build_availability_report(&mut HtmlReportBuilder::default())
    );
}
